File: Data Structures & Algorithms/subsets/submission-4.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
     List<List<Integer>> ans = new ArrayList<>();
     List<Integer> path = new ArrayList<>();
     dfs(ans, path, nums, 0);
     return ans;   
    }

    private void dfs(List<List<Integer>> ans, List<Integer> path, int[] nums, int i) {
        if (i == nums.length) {
            ans.add(new ArrayList<>(path));
            return;
        }

        // 不選
        dfs(ans, path, nums, i + 1);

        path.add(nums[i]);
        dfs(ans, path, nums, i + 1);
        path.removeLast();
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-4.py
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = [-1] * (n+1)
        memo[0] = 1
        memo[1] = 1

        def dfs(n):
            if memo[n] != -1:
                return memo[n]
            ans = dfs(n-2) + dfs(n-1)
            memo[n] = ans    
            return memo[n] 

        return dfs(n)    
